### PR TITLE
Update DTS for nrf21540

### PR DESCRIPTION
**Pull request by JakubPegza**
DTS file for nrf21540dk_52840 has been updated according to the board datasheet.

### CODE REVIEW

**Comment by ioannisg** [comment]
Would like @anangl to take a look

**Comment by hubertmis** [comment]
@AdamZelikNS, could you have a look if DTS definition is as expected?

**Comment by AdamZelikNS** [comment]
> @AdamZelikNS, could you have a look if DTS definition is as expected?

At the moment the changes made here are compatible with this [PR](https://github.com/nrfconnect/sdk-nrf/pull/2896) but that will change soon ..
I'm going to use type phandle-array (instead of int) to set some other attributes along with the pin numbers as well.
I'll let you know soon what it will look like in the end and how we will need to align this pr

**Comment by hubertmis** [comment]
We shall postpone merging it until DTS format in https://github.com/nrfconnect/sdk-nrf/pull/2896 is finalized.

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts
Comment (anangl) [inline boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:99]: This does not look like a correct change. Pins P0.00/P0.01 are LFXO pins (XL1/XL2). And those two here are supposed to  be SDA/SCL lines in the Arduino connector.

File: dts/bindings/arm/nordic,nrf21540_gpio.yaml
Comment (anangl) [inline dts/bindings/arm/nordic,nrf21540_gpio.yaml:22]: typo: `nRFf21540`

File: dts/bindings/arm/nordic,nrf21540_gpio.yaml
Comment (anangl) [inline dts/bindings/arm/nordic,nrf21540_gpio.yaml:4]: I guess it should be the Apache license, like in other such files. Also it would be good to shorten the header like it was done in other bindings.

File: dts/bindings/arm/nordic,nrf21540_gpio.yaml
Comment (anangl) [inline dts/bindings/arm/nordic,nrf21540_gpio.yaml:10]: `"nordic,nrf21540-gpio"` would probably look more in line with other `compatible` properties.

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts
Comment (JakubPegza) [reply boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:99]: In the documentation that I have, those pins are mapped to both XL1/XL2 and D14/D15. Could it be an oversight in the schematic? I'm looking at revision 0.9.0 with date 09.07.2020

Comment (hubertmis) [reply boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:99]: I've verified on revision 0.7.0 and it seems it should be `10` and `9` accordingly. Not sure about newer DKs, because I don't have any.

**Comment by hubertmis** [comment]
@JakubPegza , is it already aligned with expected binding file? Can we remove DNM label?

**Comment by JakubPegza** [comment]
@hubertmis It should be. I aligned it with PRs to NCS. @AdamZelikNS can verify if the binding is correct now and it should be good to merge.

**Comment by hubertmis** [comment]
@AdamZelikNS , please confirm if the format is as expected now

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts
Comment (anangl) [reply boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:99]: > In the documentation that I have, those pins are mapped to both XL1/XL2 and D14/D15. Could it be an oversight in the schematic? I'm looking at revision 0.9.0 with date 09.07.2020

I'm looking at the same one. Lines D14/D15 there (in Auxiliary Connector P12) are a different thing than D14/D15 here (these two correspond to SDA/SCL lines in Arduino Interface Connector P10 in the schematic).

**Comment by AdamZelikNS** [comment]
The 'nordic,nrf21540-gpio.yaml' file should also be aligned.
Please look at last commit on 'feature/KRKNWK-7379-use_phandle_array_in_FEM_dt' branch of bb ncs-test-apps repo

**Comment by AdamZelikNS** [comment]
Looks to me that it's ok now

**Review by hubertmis — APPROVED** [review]

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts
Comment (JakubPegza) [reply boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts]: It seems there was some discrepancies in pin definitions, all cleared up now.

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (KAGA164) [inline dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: I think this is not necessary. You can make `tx-en-gpios` as optional properties by changing:
```
    tx-en-gpios:
        type: phandle-array
        required: false
        description: >
            Gpio of the SOC controlling TX_EN pin of the nRF21540
```
Then in application you can check by macros if DTS has this property.

**Review by KAGA164 — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
@ppryga could you please take a look?

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (JakubPegza) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: This file should be aligned with AdamZelikNS/sdk-nrf@d217b13. @AdamZelikNS, do you think this would be a better way to handle this configuration?

Comment (AdamZelikNS) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: The idea is that the user can choose not to control one of the pins (for example `tx-en`) by removing `tx-en-gpios` property, but must declare `tx-en-gpio-is-not-used` instead. This is to better visualize why `tx-en-gpios` property is missing in such case and to prevent possible mistakes

**Comment by ioannisg** [comment]
@JakubPegza CI seems failing, pls take a look

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (hubertmis) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: With current proposal DTS would look like:

```
   / {
       nrf_radio_fem: name_of_fem_node {
           compatible = "nordic,nrf21540_gpio";
           tx-en-pin  = < &gpio0 13 ACTIVE_HIGH>;
           rx-en-gpio-is-not-used: true;
           pdn-pin    = < &gpio0 15 ACTIVE_HIGH>;
       };
   };
```

In your proposal it would be:
```
   / {
       nrf_radio_fem: name_of_fem_node {
           compatible = "nordic,nrf21540_gpio";
           tx-en-pin  = < &gpio0 13 ACTIVE_HIGH>;
           pdn-pin    = < &gpio0 15 ACTIVE_HIGH>;
       };
   };
```

It could end up with incorrect configurations created by mistake. In current binding, DTS creator has to explicitly state that given pin is not used, what should help avoiding mistakes in DTS creation.

**Comment by hubertmis** [comment]
Probably rebase is needed. I think @czeslawmakarski fixed this issue already

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (anangl) [inline dts/bindings/arm/nordic,nrf21540-gpio.yaml:7]: This should use `-` instead of `_`.
And since this binding contains not only strictly GPIO related properties but also time and gain values, maybe instead of `gpio` it should use something else in the compatible string, e.g. `fem`?

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (ppryga) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: I agree with @KAGA164. In my humble opinion this is not necessary. You can make all `xxx-en-gpios` optional. If they are set as not required then end user just does not provide particular property. Less typing for end user and less code for driver maintainer. 

BTW. I've taken a quick look into documentation. Is there a use case that not all `tx_en`, `rx_en` and `pdn` are required to control the device?

**Review by ppryga-nordic — CHANGES_REQUESTED** [review]

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (hubertmis) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:7]: nRF21540 can be controlled by two interfaces: GPIO or SPI. This binding describes controlling by GPIO, that's the reason of `_gpio` suffix. When SPI is implemented, another binding `nordic/nrf21540_spi` should be created.

**Comment by hubertmis** [comment]
@ppryga, I don't know why, but some customers require such configuration. They could connect PDN to Vcc, or try to control TX and RX with single line and NOT logic.

**Comment by ppryga** [comment]
> @ppryga, I don't know why, but some customers require such configuration. They could connect PDN to Vcc, or try to control TX and RX with single line and NOT logic.

@hubertmis ok, I was wondering if those properties have to be optional.

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (ppryga) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: @hubertmis if we go that way, by stating explicitly that particular properties are disabled, wouldn't we end with forcing end users with explicit stating that e.g. peripherals are disabled?

Comment (hubertmis) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: @ppryga , I don't have a strong opinion. I'm fine with both solutions. Just wanted to make sure that we understand the risk of mistakes if we make nRF21540 GPIO pins optional in DTS. As you noticed, these pins are __not__ optional in Product Specification.

@AdamZelikNS, @JakubPegza , what's your opinion?

Comment (JakubPegza) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: @hubertmis @ppryga I would be in favor of explicitly disabling control of specific pins, as this feels like less mistake-prone way, but I'm fine with both solutions.

Comment (ppryga) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: I would be in favor of removing the explicit disabling. It is not kind of a configuration that has internal-cross-dependencies, that we have to defend from being set inappropriately. I believe that if one starts to change it, he knows what he wants. Besides that, if pins are not optional, then I suggest to set `required=true` there. 

But I will not force it, let majority decides.  @KAGA164 ?

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts
Comment (pabigot) [inline boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts]: This is reverting a change made in the previous commit.  Either merge the two commits, or remove the change.

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (pabigot) [inline dts/bindings/arm/nordic,nrf21540-gpio.yaml:2]: This also is correcting something in the previous commit.  Probably there should just be one commit.

**Review by pabigot — COMMENTED** [review]
The two commits should be merged, or at least the second should not correct things that were done in the first.

At a high level: The general rule for devicetree in Zephyr is that devicetree describes the hardware configuration, while policy is selected by Kconfig.

On that model I don't see why this binding is specific to using the GPIO control API.  It seems to make more sense to name it `nordic,nrf21540` and provide properties that describe the assignment of pins for both GPIO and SPI capabilities.  Properties for pins that are not unconditionally required (for example `SCK` when not using SPI control) should be marked `required: false`.  They would not be present in the devicetree source if the hardware connected them to `Vss`/`Vdd` or left them floating.

All the `*-is-not-use` properties should be removed and replaced with Kconfig options, e.g. a choice between GPIO and SPI control.

The driver should verify that the policy decision from Kconfig (e.g. `CONFIG_NRF21540_CTL_GPIO=y`) is consistent with the properties provided in devicetree at build-time.

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (KAGA164) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:7]: What about antenna select pin in that case ? Where will they be configured ?

File: dts/bindings/arm/nordic,nrf21540-gpio.yaml
Comment (KAGA164) [reply dts/bindings/arm/nordic,nrf21540-gpio.yaml:16]: > They could connect PDN to Vcc, or try to control TX and RX with single line and NOT logic.

How souch cases will be handled if for example PDN pin isn't physically connected to the board and pdn pin is mandatory so user still has to pass there some GPIO port and pin ? @hubertmis @JakubPegza

**Review by pabigot — CHANGES_REQUESTED** [review]
https://github.com/zephyrproject-rtos/zephyr/pull/30046#pullrequestreview-532302574 was intended to be blocking.

**Review by KAGA164 — CHANGES_REQUESTED** [review]

**Comment by JakubPegza** [comment]
@KAGA164 We've decided to implemented changes to the devicetree incrementally, as the driver starts supporting them. As for specific pins not being required - Kconfig will prevent specific implementation (like nrf21540 gpio) from being selected if not all necessary hardware is provided by dts.

File: dts/bindings/arm/nordic,nrf21540-fem.yaml
Comment (pabigot) [inline dts/bindings/arm/nordic,nrf21540-fem.yaml:29]: There's concern in issue #25164 about use of binding defaults that might produce incorrect behavior on variant hardware.  Some proposed guidelines are [here](https://github.com/zephyrproject-rtos/zephyr/issues/25164#issuecomment-701486128).

I don't see anything in the data sheet that describes the settling and hold times, which suggests these are very much hardware-dependent.  In that case these should be made required, and the description updated to provide an example value (and perhaps mention how it can be determined).  For convenience a devicetree source fragment showing the default settings can be placed in the description of the binding itself (at the top of the file), where a user can simply copy it into the devicetree source.

If these are conservative settings that would work for any conceivable hardware platform, but might be optimized to shorter durations in certain cases, then they could remain defaults, but the description should explain how they were selected.

If they're "the values that result from using the Nordic reference design" then documenting that could be enough to justify making them defaults.

For values like gain which are presumably configurable (even if the API isn't there yet) these can be defaulted as long as the selected default value is justified in the documentation, e.g. 20 dB is the nominal value of the power-up default.

**Review by pabigot — CHANGES_REQUESTED** [review]
Thanks, this looks much better.

We need justification for making some properties default, though; see comment below.

**Comment by KAGA164** [comment]
@JakubPegza I would implement full hardware description  here including SPI interface and antenna select pin like in @pabigot comment. We have applications that need simple antenna control and can use mixed mode with spi and gpio and do not require a special driver at this moment. Could you add a description of this antenna select pin and spi interface here ?

File: dts/bindings/arm/nordic,nrf21540-fem.yaml
Comment (hubertmis) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml:29]: Please have a look at nRF21540 Product Specification: https://infocenter.nordicsemi.com/topic/struct_fem/struct/nrf21540_ps.html table 6. There are defined timings that are used as defaults in this file (I see now that our timings from OPS had different value than from PS, @JakubPegza please have a look and update wherever necessary). These timings are recommended for most designs. However, I'm aware of FEM users that tweak timings in their products. Because of that we should keep them configurable. Do you think that we should add reference to table 6 from Product Specification to the description of these fields?

Values in gain fields are default out-of-factory values. You can find them in the Product Specification as well. It is possible to modify them with SPI interface. If default values are modified with SPI interface, DTS should be modified to reflect amplification value. Should we reference to Product Specification to justify these defaults as well?

**Comment by JakubPegza** [comment]
Updated default timings to match product specification.

File: dts/bindings/arm/nordic,nrf21540-fem.yaml
Comment (pabigot) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml]: Yes, if the binding documentation includes most of the explanatory material above and references the specification that should cover it.  At least for the timings.

I would expect a SPI control driver to use the devicetree gain values to reconfigure the FEM automatically when it's initialized.  In that case I would remove the default, and leave the properties as `required: false`, then document that absence of the property means the power-up default will be used.  A compile-time diagnostic would be appropriate if the property is present but there's no way to set it.

Looking at the specification more closely I might use with properties like this:
```
    tx-gain = <22>; /* integer value for CONFREG0.TX_GAIN */
    tx-gain-poutb;   /* boolean flag selecting CONFREG0.MODE = 1 */
    pouta-sel-uicr;  /* boolean flag selecting CONFREG2.POUTA_SEL = 1 */
```
that more closely track the real interface.  If you specify gains as dB, will the driver be able to automatically translate that into a correct `TX_GAIN` register setting?

But that level of detail depends on what your users will expect.  From the Zephyr perspective we just need to know what gets a default, where it came from, and conditions under which it needs to be changed.

Comment (hubertmis) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml]: I'm thinking about this TX and RX gain values. Basically there are multiple values that could be added to DTS binding:

1. POUTA_PROD (constant = 20)
2. POUTA_UICR (optional, can be configured by board creator)
3. POUTA_SEL (bool, if PROD or UICR is being used)

and the same set for POUTB registers.

I'm not sure if there is any point in defining immutable constants like POUTA_PROD in device tree, or device tree bindings. It could be provided as a define in source file. `POUTX_UICR` and `POUTX_SEL` could be added as optional fields to DTS, where board designer could state what values are programmed in nRF21540 UICR for given board. Software could use these values to determine if MODE pin should be asserted or deasserted for given transmission.

However, I think right now power mapping, or power control modules are not implemented and MODE pin is not managed by nRF21540 driver. Because of that I think that we could remove gain-related fields from this PR and return to this discussion when these modules are implemented and could use data from these fields. @pabigot , @JakubPegza , @AdamZelikNS, what do you think?

File: dts/bindings/arm/nordic,nrf21540-fem.yaml
Comment (hubertmis) [inline dts/bindings/arm/nordic,nrf21540-fem.yaml:29]: I think 10.5 from PS should be rounded up to 11

File: dts/bindings/arm/nordic,nrf21540-fem.yaml
Comment (pabigot) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml]: The devicetree model isn't designed for named constants.  But yes, the short path is to remove the stuff you aren't ready to support and add it when there's something that would use it.

Comment (JakubPegza) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml]: I agree that if we want to add these fields incrementally, gain configuration is not necessary now. But if only `POUTX_UICR` was to be added (ant not `POUTX_PROD`), what's the use of adding `POUTX_SEL`? Selection could be done by the presence or absence of `POUTX_PROD` field then.

Comment (hubertmis) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml]: @pabigot , have I understand correctly that if we add references to nRF21540 Product Specification to timing `description` fields in the binding file and remove gain fields, you'll approve? Or do you expect any more modifications to this PR?

@JakubPegza , below a little longer explanation why I think that _SEL are needed in DTS to implement power control module.

> But if only `POUTX_UICR` was to be added (ant not `POUTX_PROD`), what's the use of adding `POUTX_SEL`?

TX power can be modified by firmware in two ways:

1. SPI control: using TX_GAIN field in CONFREG0
2. MODE pin control (when SPI is not connected from MCU do FEM)

In SPI mode it is easy: software selects any gain it wants.

But in GPIO mode it is more complicated. Software can select gain of RADIO peripheral and one of two values of gain in nRF21540. Default factory values of nRF21540 gain are POUTA_PROD (20) and POUTB_PROD(10). But these values may be replaced in factory programming UICR of nRF21540. To replace POUTA_PROD, you can program POUTA_UICR with selected value and set POUTA_SEL to 1 to start using POUTA_UICR instead of POUTB_UICR. Similarly POUTB may be replaced.

The only action firmware can do is to set MODE to 0 or to 1. When MODE is set to 0 then POUTA is used, when 1 POUTB is used. But firmware must know what gain is programmed in UICR for POUTA and POUTB to correctly select. Because of that firmware should know POUTx_UICR and POUTx_SEL to decide how MODE pin should be managed. And I think that these values could be added to DTS in the future as optional values.

Comment (pabigot) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml]: > if we add references to nRF21540 Product Specification ... you'll approve? 

Yes.  Something like:

    Default value is from Table 6 of the nRF21540 Product Specification (v1.0), 
    and can be overridden for tuned configurations.

Alternatively just remove them entirely, store the default values in the driver, and provide API in the driver to update the values under application control.  That would be the most flexible solution: maybe people will care about sub-microsecond differences, and perhaps changing based on operational state, so the user API could use `ns` values.

File: dts/bindings/arm/nordic,nrf21540-fem.yaml
Comment (mbolivar-nordic) [inline dts/bindings/arm/nordic,nrf21540-fem.yaml:7]: > We shall postpone merging it until DTS format in [nrfconnect/sdk-nrf#2896](https://github.com/nrfconnect/sdk-nrf/pull/2896) is finalized.

This PR also defines a binding for the compatible "nordic,nrf21540-fem".

May I assume that PR will be updated to remove the binding, since it is going upstream?

Comment (JakubPegza) [reply dts/bindings/arm/nordic,nrf21540-fem.yaml:7]: That PR needs to be aligned with this PR after all the changes, including removing unnecessary bindings

**Review by pabigot — APPROVED** [review]

**Review by ppryga-nordic — APPROVED** [review]

**Comment by grochu** [comment]
Hi, I have a general comment to this PR. Maybe part of the HW interface of this chip will not be used at the moment by the nRF21540 driver, but it doesn't mean it won't be used elsewhere. E.g. _PDN_, _ANT_SEL_ and _MODE_ can be controlled from a proprietary stack or from application level e.g. for testing/characterization of an RF hardware design.
@JakubPegza can this PR add the complete HW description (including the missing pins) and an SPI device in DTS? If not, we will need post another PR to this DTS description adding the missing GPIOs and go through the review over again. @pabigot, @hubertmis, what are your thoughts?

**Comment by pabigot** [comment]
> what are your thoughts?

It's not immediately obvious how this should be associated with a SPI bus: normally we would make the FEM devicetree node a child node in a SPI bus node, but that's not correct when it's not being used by SPI (when GPIO controlled it's not a bus child at all).  There are several possible solutions, but I would not try to solve that problem in this PR.

However it does seem to make sense to define all GPIO control signals now rather than submit another PR very soon that adds them.

**Comment by grochu** [comment]
> However it does seem to make sense to define all GPIO control signals now rather than submit another PR very soon that adds them.

Yes, then at least these GPIOs should be added in this PR, to avoid confusion with multiple PRs.

**Comment by hubertmis** [comment]
@grochu , ANT_SEL pin seems pretty obvious to implement.
However, I'm not sure if there is much sense in adding MODE pin without POUTx_... values. And as discussed above it is not obvious how to intorduce POUTx_... to device tree.
If you see any use case for MODE pin software usage that is not aware of POUTx_... values I'm fine to add it now. But if there is no such use case, I would delay adding MODE pin description until we decide how to describe these UICR values.

**Comment by KAGA164** [comment]
It isn't clear to me why POUTx would be presented in the devicetree ? Since by default this pin switch between two production values which are known and when SPI is interface is used it can be configured to switch between two programmable values from UICR or gain can be write directly by SPI. I think it would be good to have complete hardware pin description. Mode pin could be added, for example:
```
    mode-gpios:
        type: phandle-array
        required: false
        description: > TX power mode control
```
I don't have a use case at this moment . I'am fine also with adding mode pin later, not in this PR.

**Comment by JakubPegza** [comment]
@KAGA164 @pabigot @grochu Ant-sel GPIOs have been added now. SPI-related-gpios are left to be added in the future. MODE pin has been omitted, as there seems to be no use case for MODE without POUTx.

**Review by KAGA164 — APPROVED** [review]

**Comment by grochu** [comment]
> @KAGA164 @pabigot @grochu Ant-sel GPIOs have been added now. SPI-related-gpios are left to be added in the future. MODE pin has been omitted, as there seems to be no use case for MODE without POUTx.

Thanks. Then it seems like the chip description is complete for now. SPI should be added in a separate step.

**Review by grochu — APPROVED** [review]

**Comment by carlescufi** [comment]
@anangl and @mbolivar-nordic any objections to merging this?

**Comment by anangl** [comment]
> @anangl and @mbolivar-nordic any objections to merging this?

None from my side. It looks fine.

**Review by anangl — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**
